File: backend/packages/platform/src/migrations.ts
import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { Client } from 'pg';
import { ConfigError, DatabaseConfig, loadMigratorConfig } from './config';

export interface MigrationOptions {
  /** The service's own schema, e.g. `identity`. Its migrations may only create objects there. */
  schema: string;
  /** Folder of `NNNN_description.sql` files, applied in file-name order. */
  dir: string;
  connection: DatabaseConfig;
  log?: (message: string) => void;
}

const MIGRATION_FILE = /^\d{4}_[a-z0-9_]+\.sql$/;
const IDENTIFIER = /^[a-z_][a-z0-9_]*$/;

/**
 * Applies pending SQL migrations for one service schema, each in its own transaction, recording
 * them in `<schema>.schema_migrations`. Runs as teamora_migrator so every table is owned by it and
 * service roles never own (and so never bypass Row-Level Security on) a table.
 * A per-schema advisory lock makes concurrent runs (two deploys at once) safe.
 */
export async function runMigrations(opts: MigrationOptions): Promise<string[]> {
  if (!IDENTIFIER.test(opts.schema)) throw new Error(`invalid schema name "${opts.schema}"`);
  const log = opts.log ?? (() => undefined);
  const table = `"${opts.schema}".schema_migrations`;
  const files = readdirSync(opts.dir)
    .filter((f) => MIGRATION_FILE.test(f))
    .sort();

  const client = new Client({ ...opts.connection, application_name: `migrate:${opts.schema}` });
  await client.connect();
  try {
    await client.query('SELECT pg_advisory_lock(hashtext($1))', [`teamora-migrate:${opts.schema}`]);
    await client.query(
      `CREATE TABLE IF NOT EXISTS ${table} (version text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())`,
    );
    // Default privileges grant service roles DML on new tables; the migration history is migrator-only.
    const { rows: grantees } = await client.query<{ grantee: string }>(
      `SELECT DISTINCT grantee FROM information_schema.role_table_grants
        WHERE table_schema = $1 AND table_name = 'schema_migrations' AND grantee <> current_user`,
      [opts.schema],
    );
    for (const { grantee } of grantees) {
      if (IDENTIFIER.test(grantee) || grantee === 'PUBLIC') {
        await client.query(`REVOKE ALL ON ${table} FROM ${grantee === 'PUBLIC' ? 'PUBLIC' : `"${grantee}"`}`);
      }
    }

    const { rows } = await client.query<{ version: string }>(`SELECT version FROM ${table}`);
    const applied = new Set(rows.map((r) => r.version));
    const done: string[] = [];
    for (const file of files) {
      const version = file.replace(/\.sql$/, '');
      if (applied.has(version)) continue;
      const sql = readFileSync(join(opts.dir, file), 'utf8');
      await client.query('BEGIN');
      try {
        await client.query(sql);
        await client.query(`INSERT INTO ${table} (version) VALUES ($1)`, [version]);
        await client.query('COMMIT');
      } catch (err) {
        await client.query('ROLLBACK').catch(() => undefined);
        throw new Error(`migration ${opts.schema}/${file} failed: ${(err as Error).message}`);
      }
      log(`applied ${opts.schema}/${file}`);
      done.push(version);
    }
    if (done.length === 0) log(`${opts.schema}: up to date (${files.length} migration(s))`);
    return done;
  } finally {
    await client.query('SELECT pg_advisory_unlock_all()').catch(() => undefined);
    await client.end();
  }
}

/** Entry point for a service's `src/migrate.ts`. */
export async function runMigrationsCli(schema: string, dir: string): Promise<void> {
  try {
    const connection = loadMigratorConfig();
    await runMigrations({ schema, dir, connection, log: (m) => console.log(m) });
  } catch (err) {
    console.error(err instanceof ConfigError ? err.message : `ERROR: ${(err as Error).message}`);
    process.exit(1);
  }
}

File: backend/python-libs/teamora_common/src/teamora_common/migrations.py
"""Plain-SQL migrations, one schema per service."""

from __future__ import annotations

import re
import sys
from collections.abc import Callable
from importlib.resources import files
from importlib.resources.abc import Traversable

import psycopg

from .config import ConfigError, DatabaseConfig, load_migrator_config
from .env_file import apply_env_file_arg

_MIGRATION_FILE = re.compile(r"^\d{4}_[a-z0-9_]+\.sql$")
_IDENTIFIER = re.compile(r"^[a-z_][a-z0-9_]*$")


def run_migrations(
    schema: str,
    migrations: Traversable,
    connection: DatabaseConfig,
    log: Callable[[str], None] = lambda _: None,
) -> list[str]:
    """Applies pending ``NNNN_description.sql`` files for one service schema, in file-name order.

    Each file runs in its own transaction and is recorded in ``<schema>.schema_migrations``. Runs as
    teamora_migrator, so every table is owned by it and service roles never own (and so never bypass
    Row-Level Security on) a table. A per-schema advisory lock makes concurrent runs (two deploys at
    once) safe.
    """
    if not _IDENTIFIER.match(schema):
        raise ValueError(f'invalid schema name "{schema}"')
    table = f'"{schema}".schema_migrations'
    names = sorted(f.name for f in migrations.iterdir() if _MIGRATION_FILE.match(f.name))

    with psycopg.connect(
        host=connection.host,
        port=connection.port,
        dbname=connection.database,
        user=connection.user,
        password=connection.password,
        application_name=f"migrate:{schema}",
        autocommit=True,
    ) as conn:
        conn.execute("SELECT pg_advisory_lock(hashtext(%s))", (f"teamora-migrate:{schema}",))
        try:
            conn.execute(
                f"CREATE TABLE IF NOT EXISTS {table} (version text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())"
            )
            # Default privileges grant service roles DML on new tables; the migration history is migrator-only.
            grantees = conn.execute(
                """SELECT DISTINCT grantee FROM information_schema.role_table_grants
                    WHERE table_schema = %s AND table_name = 'schema_migrations' AND grantee <> current_user""",
                (schema,),
            ).fetchall()
            for (grantee,) in grantees:
                if grantee == "PUBLIC":
                    conn.execute(f"REVOKE ALL ON {table} FROM PUBLIC")
                elif _IDENTIFIER.match(grantee):
                    conn.execute(f'REVOKE ALL ON {table} FROM "{grantee}"')

            applied = {row[0] for row in conn.execute(f"SELECT version FROM {table}")}
            done: list[str] = []
            for name in names:
                version = name.removesuffix(".sql")
                if version in applied:
                    continue
                sql = migrations.joinpath(name).read_text(encoding="utf-8")
                try:
                    with conn.transaction():
                        # No parameters, so the whole file runs as one multi-statement script.
                        conn.execute(sql)  # type: ignore[arg-type]
                        conn.execute(f"INSERT INTO {table} (version) VALUES (%s)", (version,))
                except psycopg.Error as exc:
                    raise RuntimeError(f"migration {schema}/{name} failed: {exc}") from exc
                log(f"applied {schema}/{name}")
                done.append(version)
            if not done:
                log(f"{schema}: up to date ({len(names)} migration(s))")
            return done
        finally:
            conn.execute("SELECT pg_advisory_unlock_all()")


def migrations_cli(schema: str, package: str) -> None:
    """Entry point for ``python -m <service>.migrate [--env-file PATH]``.

    Migration files ship inside the service package, in ``<package>/migrations/``.
    """
    apply_env_file_arg(sys.argv[1:])
    try:
        run_migrations(schema, files(package).joinpath("migrations"), load_migrator_config(), log=print)
    except ConfigError as exc:
        print(exc, file=sys.stderr)
        sys.exit(1)
    except Exception as exc:  # noqa: BLE001 - one clear line for a failed deploy step
        print(f"ERROR: {exc}", file=sys.stderr)
        sys.exit(1)
